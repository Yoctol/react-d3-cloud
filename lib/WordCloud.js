'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _reactFauxDom = require('react-faux-dom');

var _reactFauxDom2 = _interopRequireDefault(_reactFauxDom);

var _d3Selection = require('d3-selection');

var _d3Scale = require('d3-scale');

var _d3Cloud = require('d3-cloud');

var _d3Cloud2 = _interopRequireDefault(_d3Cloud);

var _defaultMappers = require('./defaultMappers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ordinalScale = (0, _d3Scale.scaleOrdinal)(_d3Scale.schemeCategory10);
var defaultFillMapper = function defaultFillMapper(d, i) {
  return ordinalScale(i);
};

var defaultClickEvent = function defaultClickEvent(word) {};

var WordCloud = function (_Component) {
  _inherits(WordCloud, _Component);

  function WordCloud() {
    _classCallCheck(this, WordCloud);

    return _possibleConstructorReturn(this, (WordCloud.__proto__ || Object.getPrototypeOf(WordCloud)).apply(this, arguments));
  }

  _createClass(WordCloud, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.wordCloud = _reactFauxDom2.default.createElement('div');
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          data = _props.data,
          width = _props.width,
          height = _props.height,
          padding = _props.padding,
          font = _props.font,
          fontSizeMapper = _props.fontSizeMapper,
          fontFillMapper = _props.fontFillMapper,
          rotate = _props.rotate,
          onWordClick = _props.onWordClick;

      var wordCounts = data.map(function (text) {
        return _extends({}, text);
      });

      // clear old words
      (0, _d3Selection.select)(this.wordCloud).selectAll('*').remove();

      // render based on new data
      var layout = (0, _d3Cloud2.default)().size([width, height]).font(font).words(wordCounts).padding(padding).rotate(rotate).fontSize(fontSizeMapper).on('end', function (words) {
        (0, _d3Selection.select)(_this2.wordCloud).append('svg').attr('width', layout.size()[0]).attr('height', layout.size()[1]).append('g').attr('transform', 'translate(' + layout.size()[0] / 2 + ',' + layout.size()[1] / 2 + ')').selectAll('text').data(words).enter().append('text').style('font-size', function (d) {
          return d.size + 'px';
        }).style('font-family', font).style('fill', fontFillMapper).attr('text-anchor', 'middle').attr('transform', function (d) {
          return 'translate(' + [d.x, d.y] + ')rotate(' + d.rotate + ')';
        }).text(function (d) {
          return d.text;
        }).on('click', function (d) {
          return onWordClick(d);
        });
      });

      layout.start();

      return this.wordCloud.toReact();
    }
  }]);

  return WordCloud;
}(_react.Component);

WordCloud.defaultProps = {
  width: 700,
  height: 600,
  padding: 5,
  font: 'serif',
  fontSizeMapper: _defaultMappers.defaultFontSizeMapper,
  fontFillMapper: defaultFillMapper,
  rotate: 0,
  onWordClick: defaultClickEvent
};
exports.default = WordCloud;