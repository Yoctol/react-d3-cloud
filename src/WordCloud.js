import { Component } from 'react';
import PropTypes from 'prop-types';
import ReactFauxDom from 'react-faux-dom';
import { select } from 'd3-selection';
import { scaleOrdinal, schemeCategory10 } from 'd3-scale';
import cloud from 'd3-cloud';

import {
  defaultFontSizeMapper,
} from './defaultMappers';


const fill = scaleOrdinal(schemeCategory10);


class WordCloud extends Component {
  static propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape({
      text: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
    })).isRequired,
    width: PropTypes.number,
    height: PropTypes.number,
    padding: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.func,
    ]),
    font: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.func,
    ]),
    fontSizeMapper: PropTypes.func,
    rotate: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.func,
    ]),
    color: PropTypes.string,
    colors: PropTypes.arrayOf(PropTypes.string)
  }

  static defaultProps = {
    width: 700,
    height: 600,
    padding: 5,
    font: 'serif',
    fontSizeMapper: defaultFontSizeMapper,
    rotate: 0,
    color: undefined,
    colors: []
  }

  componentWillMount() {
    this.wordCloud = ReactFauxDom.createElement('div');
  }

  render() {
    const { data, width, height, padding, font, fontSizeMapper, rotate,
      colors, color } = this.props;
    const wordCounts = data.map(
      text => ({ ...text })
    );

    // clear old words
    select(this.wordCloud).selectAll('*').remove();

    const fillColor = (colors.length === 0 && !color)
    ? (d, i) => fill(i)
    : (d, i) => (i < colors.length ? colors[i] : color);

    // render based on new data
    const layout = cloud()
      .size([width, height])
      .font(font)
      .words(wordCounts)
      .padding(padding)
      .rotate(rotate)
      .fontSize(fontSizeMapper)
      .on('end', words => {
        select(this.wordCloud)
          .append('svg')
          .attr('width', layout.size()[0])
          .attr('height', layout.size()[1])
          .append('g')
          .attr('transform', `translate(${layout.size()[0] / 2},${layout.size()[1] / 2})`)
          .selectAll('text')
          .data(words)
          .enter()
          .append('text')
          .style('font-size', d => `${d.size}px`)
          .style('font-family', font)
          .style('fill', fillColor)
          .attr('text-anchor', 'middle')
          .attr('transform',
            d => `translate(${[d.x, d.y]})rotate(${d.rotate})`
          )
          .text(d => d.text);
      });

    layout.start();

    return this.wordCloud.toReact();
  }
}


export default WordCloud;
