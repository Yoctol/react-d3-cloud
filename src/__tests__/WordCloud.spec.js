import React from 'react';
import renderer from 'react-test-renderer';

import WordCloud from '../WordCloud';

const data = [
  { text: 'Hey', value: 100 },
  { text: 'lol', value: 200 },
  { text: 'cool', value: 1 },
  { text: 'duck', value: 10 },
];

it('should call custom fontSizeMapper', () => {
  const fontSizeMapper = jest.fn().mockReturnValue(10);
  renderer.create(<WordCloud data={data} fontSizeMapper={fontSizeMapper} />);
  expect(fontSizeMapper).toHaveBeenCalledTimes(data.length);
});

it('should call custom rotater', () => {
  const rotate = jest.fn().mockReturnValue(1);
  renderer.create(<WordCloud data={data} rotate={rotate} />);
  expect(rotate).toHaveBeenCalledTimes(data.length);
});

it('should add a custom color', () => {
  const fontFill = jest.fn().mockReturnValue('blue');
  renderer.create(<WordCloud data={data} fontFillMapper={fontFill} />);
  expect(fontFill.mock.calls.length).toBeGreaterThan(1);
});
