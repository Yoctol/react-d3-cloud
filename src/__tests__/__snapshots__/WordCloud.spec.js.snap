// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.js should contain all words 1`] = `
<div
  style={Object {}}
>
  <svg
    height={600}
    style={Object {}}
    width={700}
  >
    <g
      style={Object {}}
      transform="translate(350,300)"
    >
      <text
        style={
          Object {
            "fill": "#1f77b4",
            "fontFamily": "serif",
            "fontSize": "200px",
          }
        }
        textAnchor="middle"
        transform="translate(0,0)rotate(0)"
      >
        lol
      </text>
      <text
        style={
          Object {
            "fill": "#ff7f0e",
            "fontFamily": "serif",
            "fontSize": "10px",
          }
        }
        textAnchor="middle"
        transform="translate(-9,18)rotate(0)"
      >
        duck
      </text>
      <text
        style={
          Object {
            "fill": "#2ca02c",
            "fontFamily": "serif",
            "fontSize": "1px",
          }
        }
        textAnchor="middle"
        transform="translate(5,-16)rotate(0)"
      >
        cool
      </text>
    </g>
  </svg>
</div>
`;
